File: backend/internal/order/model/order.go
// Package model contains GORM entities for order module.
package model

import (
	"time"

	"github.com/google/uuid"

	"github.com/rajaku-printing/backend/internal/order/state"
)

type Channel string

const (
	ChannelOnline Channel = "online"
	ChannelPOS    Channel = "pos"
)

type MetodeAmbil string

const (
	MetodeAmbilPickup MetodeAmbil = "pickup"
	MetodeAmbilKirim  MetodeAmbil = "kirim"
)

type MetodeBayar string

const (
	MetodeBayarTransfer MetodeBayar = "transfer"
	MetodeBayarQRIS     MetodeBayar = "qris"
	MetodeBayarCash     MetodeBayar = "cash"
	MetodeBayarQRISPOS  MetodeBayar = "qris_pos"
)

type DesignSource string

const (
	DesignSourceUpload  DesignSource = "upload"
	DesignSourceRequest DesignSource = "request"
	// DesignSourceMixed — nilai TURUNAN untuk orders.DesignSource saat item
	// di dalamnya campur upload & request (§32.1). TIDAK PERNAH valid untuk
	// order_items.DesignSource (per item hanya upload|request, §32.5) —
	// dipakai HANYA di level order, ditulis oleh deriveDesignSource.
	DesignSourceMixed DesignSource = "mixed"
)

type DesignApprovalMode string

const (
	DesignApprovalInstantWalkin DesignApprovalMode = "instant_walkin"
	DesignApprovalAsyncNotify   DesignApprovalMode = "async_notify"
)

type Order struct {
	ID         uuid.UUID    `gorm:"type:uuid;primaryKey;default:gen_random_uuid()" json:"id"`
	Resi       string       `gorm:"uniqueIndex;size:30;not null"                   json:"resi"`
	CustomerID uuid.UUID    `gorm:"type:uuid;not null;index"                       json:"customer_id"`
	Channel    Channel      `gorm:"size:20;not null;index"                         json:"channel"`
	Status     state.Status `gorm:"size:50;not null;index"                         json:"status"`

	// Items — daftar produk dalam order ini (§32), preload terurut
	// `line_no ASC` oleh repository. Subtotal di bawah adalah Σ item.subtotal
	// (§32.2) — SATU-SATUNYA angka yang dibaca perhitungan uang (rekap,
	// invoice, struk); jangan pernah menjumlahkan Items ulang di layar uang.
	Items []OrderItem `gorm:"foreignKey:OrderID" json:"items,omitempty"`

	Subtotal int64 `gorm:"not null" json:"subtotal"`

	// Fulfillment
	MetodeAmbil            MetodeAmbil `gorm:"size:20;not null"                      json:"metode_ambil"`
	ShippingCost           *int64      `                                             json:"shipping_cost,omitempty"`
	ShippingAddress        *string     `                                             json:"shipping_address,omitempty"`
	ShippingRecipientName  *string     `gorm:"size:255"                              json:"shipping_recipient_name,omitempty"`
	ShippingRecipientPhone *string     `gorm:"size:20"                               json:"shipping_recipient_phone,omitempty"`
	ShippingCourier        *string     `gorm:"size:100"                              json:"shipping_courier,omitempty"`
	ShippingTrackingNumber *string     `gorm:"size:100"                              json:"shipping_tracking_number,omitempty"`

	// Payment
	MetodeBayar *MetodeBayar `gorm:"size:20"                              json:"metode_bayar,omitempty"`

	// Design flow — DesignSource di sini adalah TURUNAN (§32.1): upload |
	// request | mixed, dihitung oleh deriveDesignSource dari
	// order_items.DesignSource setiap kali daftar item berubah. Hanya untuk
	// ringkasan/pemilihan alur tampilan — JANGAN dipakai memvalidasi file
	// desain, itu dibaca dari OrderItem.DesignSource milik baris masing-
	// masing (§32.5). DesignBrief per item sekarang ada di OrderItem.
	DesignSource       DesignSource        `gorm:"size:20;not null"              json:"design_source"`
	DesignApprovalMode *DesignApprovalMode `gorm:"size:20"                       json:"design_approval_mode,omitempty"`

	// Discount snapshot (§28) — nilai diskon disalin ke order SAAT DIBUAT,
	// bukan cuma foreign key, supaya rekap/invoice/struk tetap benar walau
	// master diskon diubah/dihapus nanti. DiscountAmount adalah SATU-SATUNYA
	// kolom yang dipakai perhitungan uang di mana pun (rekap, invoice,
	// struk) — jangan pernah JOIN ke discounts untuk angka uang (§28.2 lapis
	// 3). Lihat migration 000027 & discountapi.Snapshot.
	DiscountID            *uuid.UUID `gorm:"type:uuid;column:discount_id"                      json:"discount_id,omitempty"`
	DiscountCodeSnapshot  *string    `gorm:"size:30;column:discount_code_snapshot"              json:"discount_code_snapshot,omitempty"`
	DiscountNameSnapshot  *string    `gorm:"size:150;column:discount_name_snapshot"             json:"discount_name_snapshot,omitempty"`
	DiscountTypeSnapshot  *string    `gorm:"size:20;column:discount_type_snapshot"              json:"discount_type_snapshot,omitempty"`
	DiscountValueSnapshot *float64   `gorm:"column:discount_value_snapshot"                     json:"discount_value_snapshot,omitempty"`
	DiscountAmount        int64      `gorm:"not null;default:0;column:discount_amount"          json:"discount_amount"`
	DiscountNote          *string    `gorm:"column:discount_note"                               json:"discount_note,omitempty"`

	// Total
	Total int64 `gorm:"not null"                               json:"total"`

	// Meta
	Notes     *string    `                                              json:"notes,omitempty"`
	CreatedBy *uuid.UUID `gorm:"type:uuid"                              json:"created_by,omitempty"`
	CreatedAt time.Time  `gorm:"not null;default:now()"                 json:"created_at"`
	UpdatedAt time.Time  `gorm:"not null;default:now()"                 json:"updated_at"`

	// Soft delete (§ super admin order tools, migration 000025). NEVER hard
	// delete an order — every reading query in repository.go MUST filter
	// `deleted_at IS NULL` so a "deleted" order stops showing up anywhere
	// (admin list, resi lookup, customer history, POS rekap, public tracking)
	// while the row + everything referencing it (payment proofs, design
	// files, invoices, state history) stays intact for audit/rekap.
	//
	// PRECISELY BECAUSE deletion makes an order invisible to POS
	// rekonsiliasi (ListPOSByDateRange also filters deleted_at IS NULL),
	// an order that's already `dibayar` or later CANNOT be soft-deleted
	// directly — state.IsDeletable(Status) gates this in both
	// service.SoftDeleteOrder (early check) and
	// repository.OrderRepository.SoftDelete (authoritative, row-locked
	// re-check). A paid order must be cancelled (→ Dibatalkan) FIRST; only
	// pre-payment statuses and Dibatalkan itself may be deleted, so a
	// deleted order was, by construction, never counted as revenue.
	DeletedAt    *time.Time `gorm:"column:deleted_at"                     json:"deleted_at,omitempty"`
	DeletedBy    *uuid.UUID `gorm:"type:uuid;column:deleted_by"           json:"deleted_by,omitempty"`
	DeleteReason *string    `gorm:"column:delete_reason"                  json:"delete_reason,omitempty"`
}

func (Order) TableName() string { return "orders" }

File: frontend/types/order.ts
// Shape response backend order module. Sinkronkan kalau backend berubah.
//
// §32 (Order Multi-Item, 31 Agustus 2026): satu order sekarang bisa memuat
// banyak baris produk (`items[]`) — kolom produk/ukuran/harga PINDAH dari
// `orders` ke `order_items`. Field UANG (subtotal, discount_amount,
// shipping_cost, total) TETAP di level order (§32.2/§28.2 lapis 3) — jangan
// pernah menjumlahkan `items[]` untuk angka itu, selalu baca field order.

export type OrderStatus =
  | 'order_masuk'
  | 'menunggu_ongkir'
  | 'menunggu_pembayaran'
  | 'menunggu_verifikasi'
  | 'dibayar'
  | 'ditolak'
  | 'desain_dikerjakan'
  | 'menunggu_approval_desain'
  | 'desain_diverifikasi'
  | 'proses_cetak'
  | 'qc'
  | 'siap_kirim'
  | 'siap_ambil'
  | 'dikirim'
  | 'selesai'
  | 'dibatalkan'

export type OrderChannel = 'online' | 'pos'
export type MetodeAmbil = 'pickup' | 'kirim'
export type MetodeBayar = 'transfer' | 'qris' | 'cash' | 'qris_pos'
/** Sumber desain PER ITEM (§32.5) — cuma dua nilai, item tidak pernah 'mixed'. */
export type DesignSource = 'upload' | 'request'
/** Turunan di level order (§32.5.2): 'mixed' kalau item-itemnya campur upload+request. */
export type OrderDesignSource = DesignSource | 'mixed'
export type DesignApprovalMode = 'instant_walkin' | 'async_notify'
export type PricingType = 'per_m2' | 'paket'

/**
 * OrderItem — satu baris produk di dalam order (§32). Field uang di sini
 * (unit_price, subtotal, discount_amount) HANYA penjelas baris — jangan
 * pernah dijumlahkan untuk angka uang order, selalu baca field level `Order`
 * (§32.2, sambungan §28.2 lapis 3).
 */
export interface OrderItem {
  /**
   * order_items.id (UUID) — dibutuhkan untuk upload file desain per item
   * (§32.5, field `order_item_id` wajib di multipart upload) dan untuk
   * koreksi baris item super admin (§32.9). Dipetakan backend sejak
   * 1 September 2026 (`orderItemResponse.ID` di
   * `backend/internal/order/handler/dto.go`) — selalu terisi untuk order
   * manapun (termasuk order lama pra-§32, lihat backfill migration 000033).
   */
  id: string
  line_no: number
  product_id?: string
  product_name: string
  material_name: string
  pricing_type: PricingType | string
  width_cm: number
  height_cm: number
  quantity: number
  unit_price: number
  subtotal: number
  discount_amount: number
  design_source: DesignSource | string
  design_brief?: string
  item_notes?: string
}

export interface Order {
  id: string
  resi: string
  status: OrderStatus | string
  channel: OrderChannel | string
  /** Baris produk (§32) — 1..20 item. Order lama (pra-migrasi) tetap tampil normal sebagai array 1 item. */
  items: OrderItem[]
  /** Level ORDER, bukan penjumlahan `items[]` (§32.2) — selalu baca ini untuk uang. */
  subtotal: number
  metode_ambil: MetodeAmbil | string
  shipping_cost?: number | null
  shipping_address?: string
  shipping_recipient_name?: string
  shipping_recipient_phone?: string
  metode_bayar?: MetodeBayar | string
  /** Turunan dari kumpulan `items[].design_source` — 'mixed' kalau campur upload+request (§32.5.2). */
  design_source: OrderDesignSource | string
  total: number
  notes?: string
  created_at: string
  /** Diskon terpakai (CLAUDE.md §28) — snapshot pada saat order dibuat, tetap tampil walau master diskon sudah dihapus. */
  discount_amount?: number | null
  discount_name_snapshot?: string | null
  discount_code_snapshot?: string | null
  discount_note?: string | null
}

export interface OrderListResponse {
  items: Order[]
  total: number
  page: number
  page_size: number
}

/**
 * Label produk siap-tampil untuk konteks ringkas (daftar order, kartu akun,
 * dst) — nama item pertama + "+N lainnya" kalau order punya >1 item.
 * Backend sudah menyediakan pola yang sama di kolom rekap (§32.8); helper ini
 * dipakai layar yang TIDAK lewat endpoint rekap (mis. `/admin/order`,
 * `/akun`) sehingga tidak menghitung ulang string itu berkali-kali di template.
 */
export function orderPrimaryProductLabel(o: Pick<Order, 'items'>): string {
  const items = o.items ?? []
  if (items.length === 0) return '—'
  const first = items[0].product_name
  return items.length > 1 ? `${first} +${items.length - 1} lainnya` : first
}
